### PR TITLE
typing ir.py - part 1

### PR DESCRIPTION
**Pull request by aorenste**
See #131852 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131845

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -2032,8 +2085,12 @@ def is_contiguous_storage_and_layout(x):
 
 
 def as_storage_and_layout(
-    x, freeze=True, want_contiguous=False, stride_order=None, allow_padding=False
-):
+    x: object,
```
Comment (eellison) [inline torch/_inductor/ir.py:2088]: should this be `IRNode` ?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -2011,15 +2064,15 @@ def reindex(index, sort_index):
...
+def is_storage_and_layout(x: object) -> bool:
     try:
         as_storage_and_layout(x, freeze=False)
         return True
     except NotImplementedError:
         return False
 
 
-def is_contiguous_storage_and_layout(x):
+def is_contiguous_storage_and_layout(x: object) -> bool:
```
Comment (eellison) [inline torch/_inductor/ir.py:2075]: IRNode ?

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -2011,15 +2064,15 @@ def reindex(index, sort_index):
         return results
 
 
-def is_storage_and_layout(x):
+def is_storage_and_layout(x: object) -> bool:
```
Comment (eellison) [inline torch/_inductor/ir.py:2067]: IRNode ?

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -2052,13 +2109,16 @@ def as_storage_and_layout(
         if freeze:
             if want_contiguous:
                 x.data.freeze_layout()
-                assert x.data.layout.is_contiguous()
+                assert (
```
Comment (eellison) [inline torch/_inductor/ir.py:2112]: if `x` is an IRNode and `get_layout` was typed as returning a Layout, you would not need this isisntance

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -2052,13 +2109,16 @@ def as_storage_and_layout(
...
+                    isinstance(x.data.layout, Layout) and x.data.layout.is_contiguous()
+                )
             elif stride_order is not None:
                 x.data.freeze_layout_with_stride_order(
                     stride_order, allow_padding=allow_padding
                 )
             else:
                 x.data.decide_layout()
+        assert isinstance(x.data.layout, Layout)
```
Comment (eellison) [inline torch/_inductor/ir.py:2121]: same here - should be unnecessary

**Review by eellison — APPROVED** [review]

File: torch/_inductor/ir.py (hunk 3 above)
Comment (Skylion007) [reply torch/_inductor/ir.py:2067]: Should this be a TypeGuard or TypeIs typing_extension?

Comment (aorenste) [reply torch/_inductor/ir.py]: The problem is that it doesn't really resolve to a type (or set of types) as much as a concept.

**Pull request closed without merging**